torch/distributed/_tensor/device_mesh.py

```diff
@@ -408,12 +414,49 @@ def all_gather(
             mesh_dim (int, optional): indicate which mesh dimension we want
                 to scatter on, we by default choose the first rank on the
                 mesh dimension as source of truth.
+            gather_dim (int, optional): Dimension to concatenate the resulting tensor.
+            gather_size (torch.Size, optional): Desired shape of the gathered tensor.
 
+            This suppport up to 1 elem of padding on gather_dim.
         Returns:
-            A :class:`Work` object
+            A :class:`torch.Tensor` object
         """
+        my_coordinate = self.get_coordinate()
```

Comment:
could we move the padding logic back to placement (sharding) types? I feel this is a dtensor sharding algorithm but not related to collective itself. A clean contract between sharding and collective would actually help other cases, and we might want to change our sharding/padding logic later as we discussed.

Reply:
Let me try that, but I believe they don't compose so the result will be worse.

Reply:
We decided to have dim gather/split separate from padding logic.